File: modules/core/app/model/result/EvaluationResult.java
package model.result;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import model.exercise.Success;

public class EvaluationResult {
  
  // FIXME: implement builder?
  
  protected Success success = Success.NONE;
  
  protected List<String> messages = new LinkedList<>();
  
  public EvaluationResult(Success theSuccess, List<String> theMessages) {
    success = theSuccess;
    messages = theMessages;
  }
  
  public EvaluationResult(Success theSuccess, String... theMessages) {
    success = theSuccess;
    Collections.addAll(messages, theMessages);
  }
  
  public static <T extends EvaluationResult> boolean allResultsSuccessful(List<T> results) {
    return results.stream().filter(EvaluationResult::isSuccessful).count() == results.size();
  }
  
  public static String concatCodeElements(List<String> elements) {
    if(elements.isEmpty())
      return "--";
    return "<code>" + String.join("</code>, <code>", elements) + "</code>";
  }
  
  public static String concatListElements(List<String> elements) {
    return "<ul><li>" + String.join("</li><li>", elements) + "</li></ul>";
  }
  
  public String getBSClass() {
    return success.getColor();
  }
  
  public String getGlyphicon() {
    switch(success) {
    case COMPLETE:
      return "glyphicon glyphicon-ok";
    case PARTIALLY:
      return "glyphicon glyphicon-question-sign";
    case NONE:
    case FAILURE:
    default:
      return "glyphicon glyphicon-remove";
    }
  }
  
  public List<String> getMessages() {
    return messages;
  }
  
  public int getPoints() {
    return success.getPoints();
  }
  
  public Success getSuccess() {
    return success;
  }
  
  public boolean isSuccessful() {
    return success == Success.COMPLETE;
  }
  
}

File: modules/core/app/model/result/EvaluationResult.scala
package model.result

import play.twirl.api.{Html, HtmlFormat}

import scala.collection.JavaConverters.asScalaBufferConverter

class EvaluationResult(val success: SuccessType = SuccessType.NONE) {
  val getBSClass: String = success.getColor

  def getGlyphicon: String = success match {
    case SuccessType.COMPLETE => "glyphicon glyphicon-ok"
    case SuccessType.PARTIALLY => "glyphicon glyphicon-question-sign"
    case (SuccessType.NONE | SuccessType.FAILURE) => "glyphicon glyphicon-remove"
  }

  def getPoints: Int = success.getPoints

  val isSuccessful: Boolean = success == SuccessType.COMPLETE
}

class CompleteResult[E <: EvaluationResult](val learnerSolution: String, val results: java.util.List[E])
  extends EvaluationResult(CompleteResult.analyzeResults(results)) {

  def renderLearnerSolution = new Html(s"<pre>${HtmlFormat.escape(learnerSolution)}</pre>")
}

object EvaluationResult {
  def allResultsSuccessful[T <: EvaluationResult](results: java.util.List[T]): Boolean = results.asScala.forall(_.isSuccessful)

  def concatCodeElements(elements: java.util.List[String]): String = elements.asScala.toList match {
    case Nil => "--"
    case els => els.map(el => s"<code>$el</code>").mkString
  }

  def concatListElements(elements: java.util.List[String]): String = elements.asScala.toList match {
    case Nil => "<ul/>"
    case els => s"<ul>${els.map(el => s"<li>$el</li>")}</ul>".mkString
  }
}

object CompleteResult {
  def analyzeResults[T <: EvaluationResult](results: java.util.List[T]): SuccessType =
    if (EvaluationResult.allResultsSuccessful(results)) SuccessType.NONE
    else SuccessType.COMPLETE
}
